### ISSUE TITLE
The standby node shutdown abnormally

### ISSUE THREAD

**Issue by HugWenjiang** [greenplum-db/gpdb#9207]
### Greenplum version or build
gp 6.1.0

### OS version and uname -a
centos 7.6.1810

### autoconf options used ( config.status --config )
max_connections -m 500 -v 1500
max_prepared_transactions -m 500 -v 500

when I set the standby node less then one day,the following occurs:
the standby node pg_log:
2019-12-10 05:19:35.601769 CST,,,p67794,th625322112,"127.0.0.1","20595",2019-12-10 05:19:35 CST,0,,,seg-1,,,,,"LOG","08P01","incomplete startup packet",,,,,,,0,,"postmaster.c",2176,
2019-12-10 05:29:35.376537 CST,,,p60466,th625322112,"127.0.0.1","20685",2019-12-10 05:29:35 CST,0,,,seg-1,,,,,"LOG","08P01","incomplete startup packet",,,,,,,0,,"postmaster.c",2176,
2019-12-10 05:39:35.407898 CST,,,p53027,th625322112,"127.0.0.1","20755",2019-12-10 05:39:35 CST,0,,,seg-1,,,,,"LOG","08P01","incomplete startup packet",,,,,,,0,,"postmaster.c",2176,
2019-12-10 05:49:35.555164 CST,,,p45267,th625322112,"127.0.0.1","20827",2019-12-10 05:49:35 CST,0,,,seg-1,,,,,"LOG","08P01","incomplete startup packet",,,,,,,0,,"postmaster.c",2176,
2019-12-10 05:59:35.863836 CST,,,p24430,th625322112,"127.0.0.1","20911",2019-12-10 05:59:35 CST,0,,,seg-1,,,,,"LOG","08P01","incomplete startup packet",,,,,,,0,,"postmaster.c",2176,
2019-12-10 06:01:06.005338 CST,,,p170965,th625322112,,,,0,,,seg-1,,,,,"FATAL","XX000","the limit of 500 distributed transactions has been reached (cdbdtxrecovery.c:571)","The global user configuration (GUC) server parameter max_prepared_
transactions controls this limit.",,,,"xlog redo distributed commit 2019-12-10 06:01:05.405184+08 gid = 646142960-0000032767 gid = 1575885304-0002113075, gxid = 2113075",,0,,"cdbdtxrecovery.c",571,"Stack trace:
1    0x9f24fc postgres errstart (elog.c:558)
2    0xad90a3 postgres redoDistributedCommitRecord (cdbdtxrecovery.c:567)
3    0x5425cf postgres xact_redo (xact.c:6380)
4    0x558692 postgres StartupXLOG (xlog.c:7273)
5    0x815943 postgres StartupProcessMain (startup.c:248)
6    0x5954b2 postgres AuxiliaryProcessMain (bootstrap.c:447)
7    0x8119bc postgres <symbol not found> (postmaster.c:5824)
8    0x815232 postgres PostmasterMain (postmaster.c:1509)
9    0x4ce641 postgres main (main.c:205)
10   0x7fc420dd03d5 libc.so.6 __libc_start_main + 0xf5
11   0x4cec1c postgres <symbol not found> + 0x4cec1c
"
2019-12-10 06:01:06.028126 CST,,,p170913,th625322112,,,,0,,,seg-1,,,,,"LOG","00000","startup process (PID 170965) exited with exit code 1",,,,,,,0,,"postmaster.c",3940,
2019-12-10 06:01:06.028183 CST,,,p170913,th625322112,,,,0,,,seg-1,,,,,"LOG","00000","terminating any other active server processes",,,,,,,0,,"postmaster.c",3679,

the master pg_log:
019-12-10 06:01:06.004041 CST,"gpadm",,p149930,th-855181184,"10.2.70.81","56786",2019-12-09 18:37:56 CST,0,,,seg-1,,,,,"LOG","08P01","unexpected EOF on standby connection",,,,,,,0,,"walsender.c",1463

**Comment by asimrp** [comment]
@HugWenjiang looks like you need to increase the `max_prepared_transactions` GUC on master as well as standby.  It is currently 500.  The error indicates that you are executing more than 500 concurrent transactions.  Which is great.  Let's increase the GUC to a suitable higher value.

```
gpconfig -c max_prepared_transactions -v 600
```

**Comment by HugWenjiang** [comment]
@asimrp  I only adjusted the max_connections and max_prepared_transactions parameters to 500 yesterday, which was 250 before.When standby is down, the master node has not experienced this problem.

**Comment by asimrp** [comment]
Hmm, a closer look at the code suggests that there is a bug.  The distributed transaction manager should maintain all transactions for which a distributed commit record has been written to WAL in a shared memory array: `shmCommittedGxactArray`.  Apparently, this array is populated only during recovery (on standby or crash recovery).  It should also be used on master under normal operation but it doesn't seem to be the case.

Still looking into this in more detail but thank you for reporting the issue @HugWenjiang

**Comment by asimrp** [comment]
It doesn't make sense to impose a limit during WAL replay on standby but not in normal operation on master.  In fact, the `shmCommittedGxactArray` does not need to be in shared memory because it is not used during normal operations and startup process is the only process accessing it during WAL replay.  It can be stored in the process-local memory of the startup process.

The maximum number of prepared transactions in normal operation is the size of `allTmGxact` array, which is
```C
uint32 TotalProcs = MaxBackends + NUM_AUXILIARY_PROCS + max_prepared_xacts;
```
The same limit should be applied during redo.  The array used to track distributed committed transactions (that are not yet forget committed) should have the same size.

**Comment by HugWenjiang** [comment]
@asimrp When can you solve the issue?The standby node goes down every few hours.

**Comment by ashwinstar** [comment]
I am looking into fixing this.

@asimrp I am afraid we will have to keep the `shmCommittedGxactArray` in shared memory. During crash recovery it's true only startup process uses it as end of recovery checkpoint is created by startup process itself. But that's not true during promotion of standby. During promotion `bgwriterLaunched` mostly would have been launched and hence `RequestCheckpoint()` will forward the request to checkpointer process. Due to this we will have to keep the structure in shared memory. Code snippet which hints to that

```C
StartupXLOG()
{
.....
		if (bgwriterLaunched)
		{
			if (fast_promote)
			{
				checkPointLoc = ControlFile->prevCheckPoint;

				/*
				 * Confirm the last checkpoint is available for us to recover
				 * from if we fail. Note that we don't check for the secondary
				 * checkpoint since that isn't available in most base backups.
				 */
				record = ReadCheckpointRecord(xlogreader, checkPointLoc, 1, false);
				if (record != NULL)
				{
					fast_promoted = true;

					/*
					 * Insert a special WAL record to mark the end of
					 * recovery, since we aren't doing a checkpoint. That
					 * means that the checkpointer process may likely be in
					 * the middle of a time-smoothed restartpoint and could
					 * continue to be for minutes after this. That sounds
					 * strange, but the effect is roughly the same and it
					 * would be stranger to try to come out of the
					 * restartpoint and then checkpoint. We request a
					 * checkpoint later anyway, just for safety.
					 */
					CreateEndOfRecoveryRecord();
				}
			}

			if (!fast_promoted)
				RequestCheckpoint(CHECKPOINT_END_OF_RECOVERY |
								  CHECKPOINT_IMMEDIATE |
								  CHECKPOINT_WAIT);
		}
		else
			CreateCheckPoint(CHECKPOINT_END_OF_RECOVERY | CHECKPOINT_IMMEDIATE);
.....
}
```
Hence, I feel best we can do is to adjust the limit of the struct to align with regular mode value.

**Comment by ashwinstar** [comment]
Still trying to trace out why the error happens, as maximum number of transactions in prepared state on a segment can only be `max_prepared_transactions`. Which means committed transactions on master and remaining to be completed on primaries can only be `max_prepared_transactions` and never more. Unless we are having weird scenarios where transactions are in prepared state on different segments (never exceeding single segment `max_prepared_transactions`) but in total it exceeds `max_prepared_transactions`. That should be rare kind of case and not a routine activity. I am surprised error is being hit every few hours.

**Comment by ashwinstar** [comment]
@HugWenjiang I am curious what kind of work-load is being run on the system? Also can you just try increasing `max_prepared_transactions` value on standby means edit the `postgresql.conf` file ONLY on standby to `600` and see if the problem goes away. Please report back if that resolves the issue.

**Comment by HugWenjiang** [comment]
@ashwinstar I tried, but the standby node is still down.Can I set max_prepared_transactions value on standby to a large value ？

**Comment by ashwinstar** [comment]
> On Feb 26, 2020, at 9:54 PM, HugWenjiang <notifications@github.com> wrote:
> 
> ﻿
> @ashwinstar I tried, but the standby node is still down.Can I set max_prepared_transactions value on standby to a large value 
> 
That’s surprising. Can you paste new error message? Large value as in? Maybe try 1000 and see what happens?

I think something else it going on which we have not fully understood. Please can you explain little about your work-load. Is it OLTP kind ?

Is it possible to have smaller repro for the issue?

**Comment by HugWenjiang** [comment]
The error:
seg-1,,,,,"FATAL","XX000","the limit of 600 distributed transactions has been reached (cdbtxrecovery.c:569)","The global user configuration (GUC) server
parameter max_prepared_transactions controls this limit.",,,,"xlog redo distributed commit 2020-02-27 13:11:29.623386+08 gid = 1935293152-0000032764 
gid = 1582351730-00006056953, gxid = 6056953",,0,,"cdbdtxrecovery.c",569,"Stack trace:
1	0x9f371c postgres errstart (elog.c:558)
2	0xadace3 postgres redoDistributedCommitRecord (cdbdtxrecovery.c:565)
3	0x542a1f postgres xact_redo (xact.c:6324)
4	0x558a72 postgres StartupXLOG (xlog.c:7283)
5	0x816303 postgres	StartupProcessMain (startup.c:248)
6	0x595902 postgres	AuxiliaryProcessMain (bootstrap.c:447)
7	0x81237c postgres <symbol not found> (postmaster.c:5824)
8	0x815bf2 postgres PostmasterMain ()postmaster.c:1509
9	0x4cec01 postgres	main (main.c:205)
10	0x7f202747f3d5 libc.so.6 __libc_start_main + 0xf5
11	0x4cf1dc postgres <symbol not found> + 0x4cf1dc
"
Our scenario is:
1. Timing acquisition count, frequency is 5 minutes
2. Timing statistics calculation

**Comment by gaos1** [comment]
as @ashwinstar mentioned above, could you try to set max_prepared_transactions to 1000 and see if it works? 

And it will be helpful if you can share a script/scenario for us to reproduce the issue including your table DDL and your queries

**Comment by HugWenjiang** [comment]
I trid to set max_prepared_transactions to 1000,but it didn't work.This issue only appeared after upgrading to gp6 and above. It did not appear at gp5 and the usage scenario is the same.

Our usage scenarios：
1. We have 90 concurrent crediting processes, Timing acquisition data, frequency is 5 minutes
2. We have 30 databases, each database will have a calculation job

**Comment by paul-guo-** [comment]
I was able to reproduce this using fault injector

Here are the steps to reproduce:
1. create a 3-node demo cluster.
2. gpconfig -c max_prepared_transactions -v 10 --skipvalidation
3. code change as below to add a fault; rebuild the binary and restart the cluster.
```
diff --git a/src/backend/cdb/cdbtm.c b/src/backend/cdb/cdbtm.c
index fbc4b1e063..1fa4fc8f5a 100644
--- a/src/backend/cdb/cdbtm.c
+++ b/src/backend/cdb/cdbtm.c
@@ -646,6 +646,8 @@ doNotifyingCommitPrepared(void)
            (errmsg("the distributed transaction 'Commit Prepared' broadcast succeeded to all the segments"),
            TM_ERRDETAIL));

+   SIMPLE_FAULT_INJECTOR("insert_forget_commit");
+
    doInsertForgetCommitted();
 }
```

3. Add a test in isolation2. Here is the sql file. You need to then add a blank answer file & add it into the schedule file.
```
  CREATE EXTENSION IF NOT EXISTS gp_inject_fault;


  create table t1 (a int); create table t2 (a int); create table t3 (a int); create table t4 (a int); create table t5 (a int);
  create table t6 (a int); create table t7 (a int); create table t8 (a int); create table t9 (a int); create table ta (a int);
  create table tb (a int); create table tc (a int); create table td (a int); create table te (a int); create table tf (a int);

  select gp_inject_fault_infinite('insert_forget_commit', 'suspend', 1);
  1&: insert into t1 values(2);
  2&: insert into t2 values(2);
  3&: insert into t3 values(2);
  4&: insert into t4 values(2);
  5&: insert into t5 values(2);

  6&: insert into t6 values(1);
  7&: insert into t7 values(1);
  8&: insert into t8 values(1);
  9&: insert into t9 values(1);
  10&: insert into ta values(1);

  11&: insert into tb values(5);
  12&: insert into tc values(5);
  13&: insert into td values(5);
  14&: insert into te values(5);
  15&: insert into tf values(5);

  1<:
```
3. Run the test and then the issue is triggered.

[pguo@host67:/data2/gpdb6/gpAux/gpdemo/datadirs/standby/pg_log]$ tail -f gpdb-2020-06-16_163544.csv

```
2020-06-16 16:36:10.419145 CST,,,p124680,th-809736256,,,,0,,,seg-1,,,,,"FATAL","XX000","the limit of 10 distributed transactions has been reached","The global user configuration (GUC) server parameter max_prepared_transactions controls this limit.","Process 124680 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.",,,"xlog redo distributed commit 2020-06-16 16:36:10.369856+08 gid = 186059200-0000032767 gid = 1592296543-0000000026, gxid = 26",,0,,"cdbdtxrecovery.c",569,"Stack trace:
1    0xae21af postgres errstart (elog.c:557)
2    0xbfb5c9 postgres redoDistributedCommitRecord (cdbdtxrecovery.c:565)
3    0x54155d postgres <symbol not found> (xact.c:6324)
4    0x5418f6 postgres xact_redo (xact.c:6452)
5    0x5595bf postgres StartupXLOG (xlog.c:7283)
6    0x8e7177 postgres StartupProcessMain (startup.c:248)
7    0x5b0432 postgres AuxiliaryProcessMain (bootstrap.c:437)
8    0x8e5f32 postgres <symbol not found> (postmaster.c:5827)
9    0x8e021d postgres PostmasterMain (postmaster.c:1500)
10   0x7d9bd3 postgres <symbol not found> (main.c:264)
11   0x3397e1ed5d libc.so.6 __libc_start_main + 0xfd
12   0x4901c9 postgres <symbol not found> + 0x4901c9
"
2020-06-16 16:38:10.527569 CST,,,p124654,th-809736256,,,,0,,,seg-1,,,,,"LOG","00000","startup process (PID 124680) exited with exit code 1",,,,,,,0,,"postmaster.c",3938,
2020-06-16 16:38:10.527631 CST,,,p124654,th-809736256,,,,0,,,seg-1,,,,,"LOG","00000","terminating any other active server processes",,,,,,,0,,"postmaster.c",3677,

```

The key point is that some 2pc query does not use full gang. A simple fix might be just removing the limit code, but I need to double confirm. And I think this issue exists on master also.

**Comment by paul-guo-** [comment]
@HugWenjiang Could you please paste some of your queries here if possible? Thus we can see if my repro is exactly same as yours.

**Comment by paul-guo-** [comment]
I created a patch to fix the issue (the "repro case") I mentioned above, see https://github.com/greenplum-db/gpdb/pull/10317. I found in this issue, max_connections is configured same as max_prepared_transactions so the repro case should not be same as the original github issue although the failing result is same.

I looked at the code this morning and can not think out the possible scenario that leads to the issue. @HugWenjiang since 6.1 is quite old and there have been a lot of change since then you might want to upgrade to see if this issue still exists.

**Comment by HugWenjiang** [comment]
@paul-guo-  When our gpdb is upgraded to 6.7.1, the problem still occurs. But after we use 6.7.1 to re-initialize the database and import the data, the problem is solved

**Comment by paul-guo-** [comment]
OK thanks. I'm closing this issue. Feel free to reopen if seeing it again.

**Comment by darthunix** [comment]
@paul-guo- this issue happened again on 6.10.1 (max_connections = 400 on master and standby, so 419 distributed transactions seems to be a correct limit). Master logs didn't have any errors or strange warnings.
```
2020-10-07 20:18:01.887107 UTC,"gpadmin","postgres",p1893,th113744000,"[local]",,2020-10-07 20:18:01 UTC,0,,,seg-1,,,,,"FATAL","57M02","the database system is in recovery mode","last replayed record at 14B/530DDAE8
- VERSION: PostgreSQL 9.4.24 (Greenplum Database 6.10.1 build 6.10.1_arenadata8-256.el7) on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39), 64-bit compiled on Sep 11 2020 18:54:28",,,,,,0,,"postmaster.c",2586,
2020-10-07 20:19:04.308739 UTC,,,p26379,th113744000,,,,0,,,seg-1,,,,,"FATAL","XX000","the limit of 419 distributed transactions has been reached","It should not happen. Temporarily increase max_connections (need postmaster reboot) on the postgres (master or standby) to work around this issue and then report a bug",,,,"xlog redo distributed commit 2020-10-07 23:19:04.055228+03; rels: base/760727/t_163730740 base/760727/t_163730739 base/760727/t_163730738 gid = 1601677675-0011224294, gxid = 11224294",,0,,"cdbdtxrecovery.c",571,"Stack trace:
1    0xbe327c postgres errstart + 0x1dc
2    0xcc8d3d postgres redoDistributedCommitRecord + 0x1fd
3    0x743f0e postgres xact_redo + 0x33e
4    0x759575 postgres StartupXLOG + 0xb15
5    0xa0de03 postgres StartupProcessMain + 0x123
6    0x7943d7 postgres AuxiliaryProcessMain + 0x527
7    0xa0d60d postgres PostmasterMain + 0x119d
8    0x6c953b postgres main + 0x1db
9    0x7f5003861505 libc.so.6 __libc_start_main + 0xf5
10   0x6d560f postgres <symbol not found> + 0x6d560f
"
2020-10-07 20:19:04.312599 UTC,,,p26355,th113744000,,,,0,,,seg-1,,,,,"LOG","00000","startup process (PID 26379) exited with exit code 1",,,,,,,0,,"postmaster.c",3948,
2020-10-07 20:19:04.312625 UTC,,,p26355,th113744000,,,,0,,,seg-1,,,,,"LOG","00000","terminating any other active server processes",,,,,,,0,,"postmaster.c",3687,
```
To my great disappointment WALs on a standby and master were lost and `Debug_print_full_dtm` GUC was off. So if this problem happens again I'll write more information here (keeping an eye on it). But it seems that there is some weird case when MaxBackends limit for global transactions is not enough (or there is some problem in WAL replaying logic when we extract WAL records with committed global transactions).

**Comment by paul-guo-** [comment]
@darthunix  While working on https://github.com/greenplum-db/gpdb/pull/10987, I found that it seems that such kind of issue happen for XLOG_XACT_DISTRIBUTED_COMMIT also. Maybe that could lead to the issue?

Also some possible further enhancement in my minds:
- maybe do not use array shmCommittedGxactArray. It seems that it is not needed to be in shared memory? So maybe use hash table or dynamic-length array instead. This might cover up such kind of issue users would be happy (we could warn instead for the case in log if possible).
- When seeing the scenario maybe dump the shmCommittedGxactArray before FATAL.
- It seems that xlogdump does not show those 2pc details in checkpoint xlog. We might want to show them.

If the theory is right, need to look at master branch code also.

Reopening the issue.

**Comment by darthunix** [comment]
@paul-guo- standby failed again trying to store `gid = 1602882527-0010383560` to `shmCommittedGxactArray`.
```
2020-10-22 09:06:34.329990 UTC,,,p20842,th-734332800,,,,0,,,seg-1,,,,,"FATAL","XX000","the limit of 419 distributed transactions has been reached","It should not happen. Temporarily increase max_connections (need postmaster reboot) on the postgres (master or standby) to work around this issue and then report a bug",,,,"xlog redo distributed commit 2020-10-22 12:06:34.089806+03 gid = 1602882527-0010383560, gxid = 10383560",,0,,"cdbdtxrecovery.c",571,"Stack trace:
1    0xbebd9c postgres errstart + 0x1dc
2    0xcd1eed postgres redoDistributedCommitRecord + 0x1fd
3    0x74c13e postgres xact_redo + 0x33e
4    0x761895 postgres StartupXLOG + 0xb15
5    0xa16603 postgres StartupProcessMain + 0x123
6    0x79c797 postgres AuxiliaryProcessMain + 0x527
7    0xa15e0d postgres PostmasterMain + 0x119d
8    0x6d167b postgres main + 0x1db
9    0x7fddd0f97505 libc.so.6 __libc_start_main + 0xf5
10   0x6dd74f postgres <symbol not found> + 0x6dd74f
"
```
But this time I was able to backup WALs on a failed standby. All WALs (except the last one) have no in-doubt global transaction records: every `gid` have `distributed commit` and `distributed forget` in WAL (according to `pg_xlogdump`). The last WAL has 11 global transactions at the end of file without `distributed forget` and `gid = 1602882527-0010383560` (from the error) is one of them. But it seems that standby failed by overflowing `max_tm_gxacts` before reading `distributed forget` for them.

The only place where we decrement `*shmNumCommittedGxacts` is `redoDistributedForgetCommitRecord()` function. I have only one explanation: there is some scenario when global transaction records stuck in `shmCommittedGxactArray` forever and with some time overflow `max_tm_gxacts` (we have a logical error in code). I've looked at https://github.com/greenplum-db/gpdb/pull/10987 but I am not sure that it can be interconnected with the current problem.

I plan to implement your ideas about replacing `shmCommittedGxactArray` with a local dynamic array and detailed warning. It seems to be a good starting point to understand the problem.

**Comment by dreamedcheng** [comment]
I also met this problem, and found the root cause. This problem raised when there are some running distributed transaction, and at the same time, a checkpoint is on fire. If the distributed transaction is at the the state of `DTX_STATE_INSERTING_FORGET_COMMITTED`, but actually its COMMIT FORGET xlog was written to disk. At this moment, the checkpoint process is calculating the the committed transactions in `getDtxCheckPointInfo`, it will include the distributed transaction to its checkpoint xlog, because no lock to protect the DtxState for this concurrent update and read. So,  when standby relay the xlog of checkpoint, it will add the already forgotten DTX to shmCommittedGxactArray again, which will increase the size of `shmCommittedGxactArray`  continuously until overflow the max_tm_gxacts. 

To solve this problem, I think this no need to replay the xlog of DTX on standby, so we can just skip this step on standby to fix the issue. Correct me if am wrong. And if this is the right solution, I am pleasure to issue a PR.

**Comment by paul-guo-** [comment]
> I also met this problem, and found the root cause. This problem raised when there are some running distributed transaction, and at the same time, a checkpoint is on fire. If the distributed transaction is at the the state of `DTX_STATE_INSERTING_FORGET_COMMITTED`, but actually its COMMIT FORGET xlog was written to disk. At this moment, the checkpoint process is calculating the the committed transactions in `getDtxCheckPointInfo`, it will include the distributed transaction to its checkpoint xlog, because no lock to protect the DtxState for this concurrent update and read. So, when standby replay the xlog of checkpoint, it will add the already forgotten DTX to shmCommittedGxactArray again, which will increase the size of `shmCommittedGxactArray` continuously until overflow the max_tm_gxacts.
> 
> To solve this problem, I think this no need to replay the xlog of DTX on standby, so we can just skip this step on standby to fix the issue. Correct me if am wrong. And if this is the right solution, I am glad to issue a PR.

Thanks, feel free to open a PR.  I agree with you but you might want to use gdb or write a simple isolation2 test at first to verify the theory. Replaying COMMIT FORGET is still needed. Considering the case of coordinator failover, after standby is promoted, new coordinator needs to make sure the 2PC commit finishes on segments, e.g. the below scenario: coordinator firstly write commit locally and then panic/poweroff before dispatching the prepared commit on segments, standby is promoted...

@darthunix  I meant #10987 is due to race in checkpoint related to PREPARE, but race in checkpoint related to DISTRIBUTED COMMIT  should exist also.

To fix this issue maybe move
```
getDtxCheckPointInfo(&dtxCheckPointInfo, &dtxCheckPointInfoSize);
```
just before
```
XLogRegisterData((char *) dtxCheckPointInfo, dtxCheckPointInfoSize);
```
I'm not 100% sure. Should find the history and check the code see there is any tricky reason that getDtxCheckPointInfo() is called earlier before the checkpoint xlog write log, also I suspect that the variable `includeInCkpt` is not needed anymore (please double confirm).

**Comment by dreamedcheng** [comment]
@paul-guo- I mean when promote the standby, we can revise some of its current logic. 

1. The standby don't need replay the DTX xlog.
2. when standby is going to promote, we force the startup process redo the XLOG begin at the last checkpoint's redo point. So the DTX recover process can recover the in-going DTX. 

This fix may be a little bit complicated, but I think it can help enhance the performance by the way, because the standby don't need replay DTX xlog anymore.

**Comment by dreamedcheng** [comment]
Another fix is more straightforward, when replaying checkpoint xlog at standby, we first check if gid in `shmCommittedGxactArray` can be found in checkpoint's `gxact_checkpoint->committedGxactArray`, if can not be found, we can safely remove the gid form `shmCommittedGxactArray`. After this check, we add `gxact_checkpoint->committedGxactArray` to `shmCommittedGxactArray as before. In this way, we can remove redundant gid in  `shmCommittedGxactArray`.

**Comment by darthunix** [comment]
@dreamedcheng thanks for your explanation. I have manually reproduced the problem (gdb to startup process on a mirror + three psql sessions):
```sql
1: create table t (a int);
1: select gp_inject_fault_infinite('dtm_before_insert_forget_comitted', 'suspend', dbid)
from gp_segment_configuration where role = 'p' and content = -1;
1: select gp_inject_fault('checkpoint_after_redo_calculated', 'suspend', dbid)
from gp_segment_configuration where role = 'p' and content = -1;
1: select gp_wait_until_triggered_fault('dtm_before_insert_forget_comitted', 1, dbid)
from gp_segment_configuration where role = 'p' and content = -1;

2: insert into t select generate_series(1, 10);

1: select gp_wait_until_triggered_fault('checkpoint_after_redo_calculated', 1, dbid)
from gp_segment_configuration where role = 'p' and content = -1;

3: checkpoint;

1: select gp_inject_fault_infinite('dtm_before_insert_forget_comitted', 'reset', dbid)
from gp_segment_configuration where role = 'p' and content = -1;
1: select gp_inject_fault_infinite('checkpoint_after_redo_calculated', 'reset', dbid)
from gp_segment_configuration where role = 'p' and content = -1;
```
gdb had a breakpoint on `redoDistributedCommitRecord` function and it shows out that we don't free `shmCommittedGxactArray` and decrement `*shmNumCommittedGxacts` correctly.

**Comment by darthunix** [comment]
Here is the problem
```
QD: commit (gids: 1, 2)
CP: getDtxCheckPointInfo (gids: 1, 2)
QD: forget (gids: 2) - gid 1 is still in progress
CP: XLogInsert (gids: 1, 2 are not forgotten, so gid 2 again added to shmCommittedGxactArray)
```
I think we should fix inconsistency  on a `getDtxCheckPointInfo` step - may be add a lock protection to `TMGXACT_LOG` elements of `shmCommittedGxactArray`? I can also make a PR if this idea seems good.

**Comment by darthunix** [comment]
@paul-guo- am 
1. If we want to move `getDtxCheckPointInfo` just to `XLogRegisterData` it should be done in a critical section?
2. Should we protect `TMGXACT_LOG` elements of `shmCommittedGxactArray` with locks?

@dreamedcheng 
After step `QD: forget` we don't have a `gid` 2 but according to your algorithm we'll again have to add it when read committed checkpoint message from WAL (or may be I didn't get the idea?)

**Comment by dreamedcheng** [comment]
> @paul-guo- am
> 
> 1. If we want to move `getDtxCheckPointInfo` just to `XLogRegisterData` it should be done in a critical section?
> 2. Should we protect `TMGXACT_LOG` elements of `shmCommittedGxactArray` with locks?
> 
> @dreamedcheng
> After step `QD: forget` we don't have a `gid` 2 but according to your algorithm we'll again have to add it when read committed checkpoint message from WAL (or may be I didn't get the idea?)

@darthunix checkpoint XLOG must have all running DTXs' messages at that point it generates, so just we need use the latest checkpoint XLOG to add the gid it contained to `shmCommittedGxactArray`, and the gids of `shmCommittedGxactArray` already contained can be cleared.

**Comment by darthunix** [comment]
@dreamedcheng you say we can clear already contained gids and replace `shmCommittedGxactArray` with gids from checkpoint (CP). What should we do in the situation below?
```
                    getDtxCheckPointInfo                                     XLogRegisterData +
                    (gid#1)                                                  XLogInsert (gid#1 only)
          commit1                           forget1    commit2    forget2
          +---------+-----------------------+----------+----------+----------+
          ^         ^                       ^          ^          ^          ^
          |         |                       |          |          |          |
gid#1     +         |                       +          |          |          |
                    |                                  |          |          |
                    |                                  |          |          |
gid#2               |                                  +          +          |
                    |                                                        |
                    |                                                        |
checkpoint          +                                                        +

```

**Comment by darthunix** [comment]
I have an argument why we can't replace `shmCommittedGxactArray` - in a scenario below we'll loose gid 2 and have only a wrong gid 1.
```
                                                                                                    CRASH
                    getDtxCheckPointInfo                                     XLogRegisterData +
                    (gid#1)                                                  XLogInsert (gid#1 only)
          commit1                           forget1    commit2                                                forget2
          +---------+-----------------------+----------+---------------------+----------------------+---------+
          ^         ^                       ^          ^                     ^                      ^         ^
          |         |                       |          |                     |                      |         |
gid#1     +         |                       +          |                     |                      |         |
                    |                                  |                     |                      |         |
                    |                                  |                     |                      |         |
gid#2               |                                  +                     |                      |         +
                    |                                                        |                      |
                    |                                                        |                      |
checkpoint          +                                                        +                      |
                                                                                                    |
                                                                                                    |
                                                                                                    |
                                                                                                    |
                                                                                                    |
                                                                                                    +

```
I think we should block any other WAL writes when checkpoint is performing `getDtxCheckPointInfo` - `XLogInsert` code.

**Comment by dreamedcheng** [comment]
Yes, and this just the case that incur the overflow problem. In my fix, the gid#1 remained after the current CP XLOG replay will be cleared by the next CP XLOG's replay.  So the standby will not overflow the `shmCommittedGxactArray`.

**Comment by darthunix** [comment]
But it is still possible to have an overflow in a case when MaxBackends global transactions are not forgotten yet and there is some extra gids arrived from the last checkpoint.

**Comment by dreamedcheng** [comment]
The extreme case is 2* MaxBackends, so we should ensure that max_tm_gxacts is lager than 2* MaxBackends.

**Comment by darthunix** [comment]
And one more thing. On the last diagram we still loose gid 2. Chackpointer took a gid snapshot when gid 1 was committed and there was no gid 2 at all. After that gid 2 arrived to WAL log file. Checkpointer wrote a commit message with gid 1 and it would be the starting point to replay WALs on recovery. So, when we crash right after checkpoint's commit (before gid 2 have been forgotten) we don't try to validate in-doubt gid 2 on segments (and it seems to be a problem)

**Comment by dreamedcheng** [comment]
Thanks.
It seems that the fix can't handle this case. What if we re-add the forgotten gid to `shmCommittedGxactArray` if we cannot find commit gid in it, I am not sure if there is any other case it can't handle.

**Comment by dreamedcheng** [comment]
And I have a new idea, when the standby startup, it replays the first checkpoint XLOG it received, and skip the following checkpoint XLOG's DTXs info it received, because recording DTX info in checkpoint XLOG is for the case DTX recovery.
Correct me if I am wrong.

**Comment by paul-guo-** [comment]
> And one more thing. On the last diagram we still loose gid 2. Chackpointer took a gid snapshot when gid 1 was committed and there was no gid 2 at all. After that gid 2 arrived to WAL log file. Checkpointer wrote a commit message with gid 1 and it would be the starting point to replay WALs on recovery. So, when we crash right after checkpoint's commit (before gid 2 have been forgotten) we don't try to validate in-doubt gid 2 on segments (and it seems to be a problem)

No worry for this case, gid2 XLOG_XACT_DISTRIBUTED_COMMIT redo (it's behind the checkpoint.redo LSN because getDtxCheckPointInfo() is called after setting checkpoint.redo) function xact_redo_distributed_commit() would add the gid into shmCommittedGxactArray[].

Using locking seems to be heavy since this affects the OLTP performance much. 

@dreamedcheng  Your fix solution:
"when standby is going to promote, we force the startup process redo the XLOG begin at the last checkpoint's redo point. So the DTX recover process can recover the in-going DTXs.". Re-redoing those xlogs is tricky. For your solution 2 I do not understand.

Also please ignore my previous wrong guess of  the solution that moves `getDtxCheckPointInfo` later. This won't fix the issue.

`2* MaxBackends` limitation seems to be correct so setting max_tm_gxacts fixes the overflow issue but those false positive cases still slow down the dtx recovery. Maybe on standby among two checkpoint we logged all distributed committed gid even it is forgotten and then when redoing checkpiont we removes those false positive cases?

**Comment by darthunix** [comment]
@paul-guo- if we move [getDtxCheckPointInfo](https://github.com/greenplum-db/gpdb/blob/76c99759f38fadac438cf0a8b6a2ff08d4fc1c03/src/backend/access/transam/xlog.c#L9101) just near [XLogRegisterData](https://github.com/greenplum-db/gpdb/blob/76c99759f38fadac438cf0a8b6a2ff08d4fc1c03/src/backend/access/transam/xlog.c#L9213) + [XLogInsert](https://github.com/greenplum-db/gpdb/blob/76c99759f38fadac438cf0a8b6a2ff08d4fc1c03/src/backend/access/transam/xlog.c#L9215) and wrap these instructions with `WALInsertLock` it should not affect OLTP much. And it should protect us from `forget` commits invalidating `getDtxCheckPointInfo` global transaction snapshot. Or I miss something?

**Comment by darthunix** [comment]
> No worry for this case, gid2 XLOG_XACT_DISTRIBUTED_COMMIT redo (it's behind the checkpoint.redo LSN because getDtxCheckPointInfo() is called after setting checkpoint.redo) function xact_redo_distributed_commit() would add the gid into shmCommittedGxactArray[].

@paul-guo- is it correct for the case when master crashes? (I agree that we don't have problems on a standby as it always consumes WAL records with redo less than checkpoint redo). But in a case of a crash between a checkpoint `commit` and gid 2 `forget` master starts its replay from the last checkpoint from `pg_control`. And this checkpoint contains only gid 1 but doesn't contain gid 2. And we can't be sure that master's memory is not corrupted, so we have to replace `shmCommittedGxactArray` on master with information from checkpoint `commit` record.

**Comment by darthunix** [comment]
@paul-guo- finally I got what you were talking about - I didn't understood REDO meaning in a checkpoint. Sorry, my suggestions were horrible.
```
                          getDtxCheckPointInfo                                   Checkpoint
REDO        commit 1              (gid 1)            forget 1      commit 2      (gid 1)      forget 2
  +-----------+---------------------+-------------------+-------------+-------------+------------+
              ^                     ^                   ^             ^             ^            ^
              |                     |                   |             |             |            |
              |                     |                   |             |             |            |
              |                     |                   |             |             |            |
              +                     +                   +             +             +            +
```
May be we should keep in local memory a dynamic array of `forget` gids since last checkpoint's `REDO` (gid 1 on a schema above). When we meet a new checkpoint record, we extract all gids from checkpoint (gid 1), exclude all `forget` gids from dynamic array (gid 1) and flush `forget` dynamic array. Finally we should add filtered gids (none) that are absent in `shmCommittedGxactArray`(gid 2) to it. We still need `MaxConnections` capacity for `shmCommittedGxactArray` to keep global transactions.

**Comment by paul-guo-** [comment]
I just traced the code. Here are some tentative findings.

1. checkpoint xlog rm_redo function and restartpoint code does not handle the two phase transaction list (Prepared, distributed committed-but-not-forgotten). See `xlog_redo()` and `CreateRestartPoint`

2. The issue is as below, right?
```
commit gid 1 <- or after checkpoint.redo but before getDtxCheckPointInfo()
checkpoint.redo
getDtxCheckPointInfo() get gid1 for checkpoint xlog
forget gid 1
checkpoint  <- when redoing it on standby, gid 1 is added to the distribution transaction list again.
```

2. But  in startup redo loop, checkpoint xlog is additionally handled to update the distribution transaction list (on standby only), see code,
```
                  if (IsStandbyMode() &&
                      record->xl_rmid == RM_XLOG_ID &&
                      (xlogRecInfo == XLOG_CHECKPOINT_SHUTDOWN
                       || xlogRecInfo == XLOG_CHECKPOINT_ONLINE))
                  {
                      XLogProcessCheckpointRecord(record);
                      memcpy(&checkPoint, XLogRecGetData(record), sizeof(CheckPoint));
                  }
```

It seems that we do not need the code here since in startup code, initially ReadCheckpointRecord() is called to populate the transaction list and then in the subsequent xlog redo loop we could simply depend on those 2pc xlogs to keep the list updated. I have not done experiments to verify the theory but if the theory is right we could fix this by simply removing these lines. Also it seems if apply the theory to prepared transaction, orphaned prepared transaction might be wrongly included also?

**Comment by darthunix** [comment]
1. I am not sure about the meaning of restore point. AFAICS from the code it is done by checkpointer when pg instance is in recovery mode (by the same schedule as common checkpoints are done in normal mode). Restore point updates pg_control file with the last processed checkpoint from the WAL and shrink WALs prior to last checkpoint's REDO.
So if recovery fails it starts from the last checkpoint record in WAL containing a list of global transactions. Do we need some special handle for global transactions here?
Also I didn't get the idea about [rm_redo](https://github.com/greenplum-db/gpdb/blob/54451fc061af2e2d1ba592ed962f58f73a5058ea/src/backend/access/transam/xlog.c#L7420). Under hood it is implemented with `xlog_redo()`. When we do recovery and extract a new checkpoint record from WAL, `xlog_redo()` simply runs `RecoveryRestartPoint()` to update last checkpoint information for the next `CreateRestartPoint()`. Do we miss here something?
2. Yes, the issue is described correctly.
3. Interesting fact: code works correctly on [5X_STABLE](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/access/transam/xlog.c#L6952) thanks to [getTmLock() ](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/cdb/cdbtm.c#L1829) used by [doInsertForgetCommitted()](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/cdb/cdbtm.c#L692) and [getDtxCheckPointInfoAndLock()](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/cdb/cdbtm.c#L2773) functions. So it is safe in 5X_STABLE to add all global transactions from any checkpoint record to `shmCommittedGxactArray` if they are still not there. But to tell the truth I also don't understand why should we do it for every checkpoint on a standby rather that extract it only once when we start a recovery and apply `commit` + `forget` records  to `shmCommittedGxactArray` since checkpoint's REDO.
Since 6X_STABLE we got rid of `getTmLock()` for performance reasons but didn't handle the race condition between checkpointer and QD (the problem we have now).
I plan to remove this code and make experiments.
> Also it seems if apply the theory to prepared transaction, orphaned (actually committed or aborted already) prepared transaction might be wrongly included in the list also?

I am not sure that understand this question right, but `getDtxCheckPointInfo()` grabs only `shmCommittedGxactArray` transactions (that are committing at the moment, not prepared ones)

**Comment by paul-guo-** [comment]
>     1. I am not sure about the meaning of restore point. AFAICS from the code it is done by checkpointer when pg instance is in recovery mode (by the same schedule as common checkpoints are done in normal mode). Restore point updates pg_control file with the last processed checkpoint from the WAL and shrink WALs prior to last checkpoint's REDO.
>        So if recovery fails it starts from the last checkpoint record in WAL containing a list of global transactions. Do we need some special handle for global transactions here?
>        Also I didn't get the idea about [rm_redo](https://github.com/greenplum-db/gpdb/blob/54451fc061af2e2d1ba592ed962f58f73a5058ea/src/backend/access/transam/xlog.c#L7420). Under hood it is implemented with `xlog_redo()`. When we do recovery and extract a new checkpoint record from WAL, `xlog_redo()` simply runs `RecoveryRestartPoint()` to update last checkpoint information for the next `CreateRestartPoint()`. Do we miss here something?

This is not relevant of restore point. It's called normally on standby in xlog redo loop.

```
pguo       1053    963  0 13:43 ?        00:00:00 postgres:  6001, startup process   recovering 000000010000000000000003

[pguo@host67@6:/data2]$ gdb -p 1053

(gdb) bt
#0  XLogProcessCheckpointRecord (rec=0x2880f60) at xlog.c:6142
#1  0x000000000055a44e in StartupXLOG () at xlog.c:7221
#2  0x00000000008ea1c3 in StartupProcessMain () at startup.c:242
#3  0x00000000005b162d in AuxiliaryProcessMain (argc=2, argv=0x7fff1a087820) at bootstrap.c:438
#4  0x00000000008e8f7d in StartChildProcess (type=StartupProcess) at postmaster.c:5837
#5  0x00000000008e3211 in PostmasterMain (argc=6, argv=0x283d920) at postmaster.c:1510
#6  0x00000000007dc54b in main (argc=6, argv=0x283d920) at main.c:245
```

>     2. Yes, the issue is described correctly.
> 
>     3. Interesting fact: code works correctly on [5X_STABLE](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/access/transam/xlog.c#L6952) thanks to [getTmLock() ](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/cdb/cdbtm.c#L1829) used by [doInsertForgetCommitted()](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/cdb/cdbtm.c#L692) and [getDtxCheckPointInfoAndLock()](https://github.com/greenplum-db/gpdb/blob/9b690a9e936e0bbfdbfd1b1181b39f6b92c9e2bb/src/backend/cdb/cdbtm.c#L2773) functions. So it is safe in 5X_STABLE to add all global transactions from any checkpoint record to `shmCommittedGxactArray` if they are still not there. But to tell the truth I also don't understand why should we do it for every checkpoint on a standby rather that extract it only once when we start a recovery and apply `commit` + `forget` records  to `shmCommittedGxactArray` since checkpoint's REDO.
>        Since 6X_STABLE we got rid of `getTmLock()` for performance reasons but didn't handle the race condition between checkpointer and QD (the problem we have now).
>        I plan to remove this code and make experiments.
> 
> 
> > Also it seems if apply the theory to prepared transaction, orphaned (actually committed or aborted already) prepared transaction might be wrongly included in the list also?
> 
> I am not sure that understand this question right, but `getDtxCheckPointInfo()` grabs only `shmCommittedGxactArray` transactions (that are committing at the moment, not prepared ones)

I mean
```
prepreard gxact 1
checkpoint.redo
committed gxact 1
checkpoint (redo will includes gxact1 again)
```
Looked at the code further. `StartupXLOG()->XLogProcessCheckpointRecord()` does not add prepared transactions so this is not an issue for prepared transaction.

**Comment by dreamedcheng** [comment]
@paul-guo- @darthunix 
I open a PR for this issue. Could you have a look at it?

**Comment by paul-guo-** [comment]
The fix is  in master and gpdb 6 now. Closing this issue.
